File: problem_005.rs
// -*- compile-command: "rustc -o problem_005_rs problem_005.rs" -*-
// Copyright (c) 2016 Michael Caldwell

fn is_prime(n: u64) -> bool {
    if n < 2 {
        return false;
    }
    if n == 2 {
        return true;
    }
    if n % 2 == 0 {
        return false;
    }

    let limit = (n as f64).sqrt() as u64;
    for i in 3..(limit+1) {
        if n % i == 0 {
            return false;
        }
    }
    return true;
}

fn main() {
    println!("Project Euler - Problem 5");
    println!("What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?\n");

    // Start off with a base of all primes less than 20
    // Impossible for number to be smaller than this
    let mut n = 1;
    for i in 1..20 {
        if is_prime(i) == true {
            n *= i;
        }
    }

    let mut valid = false;
    while valid == false {
        n += 10;
        valid = true;
        // No need to check 2 - 10,
        // because multiples of each are tested in range of 11-20
        for i in 11..21 {
            if n % i != 0 {
                valid = false;
                break;
            }
        }
    }

    println!("{}", n);
}

File: problem_005.go
// -*- compile-command: "go build -o problem_005_go problem_005.go" -*-
// Copyright (c) 2017 Michael Caldwell
package main

import "fmt"

func isPrime(n int) bool {
	if n < 2 { return false }
	if n == 2 { return true }
	if n % 2 == 0 { return false }

	for i := 3; i*i <= n; i += 2 {
		if n % i == 0 {
			return false
		}
	}

	return true
}

func divisible(n int) bool {
	for i := 1; i <= 20; i++ {
		if n % i != 0 {
			return false
		}
	}

	return true
}

func main() {
        fmt.Println("Project Euler - Problem 5:")
        fmt.Println("What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?\n")

	n := 1
	limit := 20

	for i := 1; i < limit; i++ {
		if isPrime(i) {
			n *= i
		}
	}

	
	for i := 2; i < limit; i++ {
		n *= i
		if divisible(n) == true {
			break
		}
	}

        fmt.Println(n)
}

